# 1.Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from env import host, user_name, password

def get_db_url(username,password,host,db_name):
    return f'mysql+pymysql://{user_name}:{password}@{host}/{db_name}'
    
url =  get_db_url(user_name,password,host, 'employees')

# 2.Use your function to obtain a connection to the employees database.

In [3]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

In [4]:
query = '''
SELECT
    t.title as title,
    d.dept_name as dept_name
FROM titles t
JOIN dept_emp USING (emp_no)
JOIN departments d USING (dept_no)
LIMIT 100
'''

title_dept = pd.read_sql(query, url)
title_dept.head()

title         dept_name
0         Staff  Customer Service
1  Senior Staff  Customer Service
2         Staff  Customer Service
3  Senior Staff  Customer Service
4         Staff  Customer Service

# 3.Once you have successfully run a query:
a. Intentionally make a typo in the database url. What kind of error message do you see?

b. Intentionally make an error in your SQL query. What does the error message look like?

In [5]:
url = get_db_url(hofst, username, password)
pd.read_sql('SELECT * FROM employtees LIMIT 5 OFFSET 50', url)

NameError: name 'hofst' is not defined

# 4.Read the employees and titles tables into two separate DataFrames.

In [8]:
emp_df = pd.read_sql('SELECT * FROM employees ', url)
emp_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [10]:
titles_df = pd.read_sql('SELECT * FROM titles', url)
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

# 5.How many rows and columns do you have in each DataFrame? Is that what you expected?

For employees table, it has 300024 rows × 6 columns

For titles table, it has 443308 rows × 4 columns

# 6.Display the summary statistics for each DataFrame.

In [11]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

In [12]:
emp_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

# 7.How many unique titles are in the titles DataFrame?

In [16]:
titles_df.title.describe()
# or
titles_df["title"].nunique()

count       443308
unique           7
top       Engineer
freq        115003
Name: title, dtype: object

# 8.What is the oldest date in the to_date column?

In [19]:
titles_df.to_date.sort_values().head(1)
titles_df["to_date"].min() #other way

datetime.date(1985, 3, 1)

# 9.What is the most recent date in the to_date column

In [22]:
titles_df.to_date.sort_values().tail(1)
titles_df["to_date"].max() #other way

datetime.date(9999, 1, 1)